File: substrate/settings_model_driver_tab_compose.py
"""Settings model driver tab compose (pure, advisory).

Decision-tree model selection with usage bar, projection, optional
Antiek-bench best, and NotDiamond shadow (never authority).

live_router_authorized always False.
secrets_stored always False.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from typing import Any

from substrate.model_decision.prompt_compose import (
    ModelDecisionPromptComposeError,
    ModelDecisionPromptComposeResult,
    compose_model_decision_with_projection,
)

_SECRETISH = re.compile(r"sk-|api[_-]?key|secret|bearer\s", re.I)


class SettingsModelDriverTabComposeError(ValueError):
    """Fail-closed validation for settings model driver tab compose."""


@dataclass(frozen=True)
class SettingsModelDriverTabCompose:
    decision: ModelDecisionPromptComposeResult
    inventory_count: int
    pending_add_count: int
    bench_aligned: bool | None
    bench_best_for_focus: str | None
    nd_shadow_differs: bool | None
    nd_shadow_model: str | None
    tab_ready: bool
    live_router_authorized: bool
    secrets_stored: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "decision": self.decision.to_dict(),
            "inventory_count": self.inventory_count,
            "pending_add_count": self.pending_add_count,
            "bench_aligned": self.bench_aligned,
            "bench_best_for_focus": self.bench_best_for_focus,
            "nd_shadow_differs": self.nd_shadow_differs,
            "nd_shadow_model": self.nd_shadow_model,
            "tab_ready": self.tab_ready,
            "live_router_authorized": False,
            "secrets_stored": False,
            "notes": list(self.notes),
            "authority": "settings_model_driver_tab_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise SettingsModelDriverTabComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_settings_model_driver_tab(
    *,
    selected_model_id: object,
    models: object,
    daily_cap_usd: object,
    spent_usd: object,
    projected_cost_usd_high: object | None = None,
    projected_cost_usd_low: object | None = None,
    bench_bests: object | None = None,
    focus_task: object | None = None,
    nd_shadow: object | None = None,
    pending_add_model_ids: object | None = None,
) -> SettingsModelDriverTabCompose:
    """Compose settings model-driver tab snapshot. Never auto-routes."""
    notes: list[str] = [
        "live_router_authorized=false — operator selects model; no auto-router",
        "secrets_stored=false — inventory/ids only; never accept raw API keys here",
        "NotDiamond is advisory/shadow only (§16 REJECT as production router)",
    ]

    try:
        decision = compose_model_decision_with_projection(
            selected_model_id=selected_model_id,
            models=models,
            daily_cap_usd=daily_cap_usd,  # type: ignore[arg-type]
            spent_usd=spent_usd,  # type: ignore[arg-type]
            projected_cost_usd_high=projected_cost_usd_high,  # type: ignore[arg-type]
            projected_cost_usd_low=projected_cost_usd_low,  # type: ignore[arg-type]
        )
    except ModelDecisionPromptComposeError as e:
        raise SettingsModelDriverTabComposeError(str(e)) from e

    notes.extend(decision.notes)
    inventory_count = len(models) if isinstance(models, list) else 0
    notes.append(f"inventory_count={inventory_count}")

    pending_add_count = 0
    if pending_add_model_ids is not None:
        if not isinstance(pending_add_model_ids, list):
            raise SettingsModelDriverTabComposeError(
                "pending_add_model_ids must be an array when set"
            )
        seen: set[str] = set()
        for i, raw in enumerate(pending_add_model_ids):
            mid = _require_nonempty(
                raw, field=f"pending_add_model_ids[{i}]"
            )
            if len(mid) > 128 or _SECRETISH.search(mid) or " " in mid:
                raise SettingsModelDriverTabComposeError(
                    f"pending_add_model_ids[{i}] must be a model id, not secret material"
                )
            if mid in seen:
                raise SettingsModelDriverTabComposeError(
                    f"duplicate pending_add_model_id: {mid}"
                )
            seen.add(mid)
        pending_add_count = len(seen)
        notes.append(
            f"pending_add_count={pending_add_count} (ids only, no secrets)"
        )

    bench_aligned: bool | None = None
    bench_best_for_focus: str | None = None
    if bench_bests is not None:
        if not isinstance(bench_bests, list):
            raise SettingsModelDriverTabComposeError(
                "bench_bests must be an array when set"
            )
        focus: str | None = None
        if focus_task is not None:
            focus = _require_nonempty(focus_task, field="focus_task")
        if len(bench_bests) == 0:
            notes.append("bench_bests empty — no invent leaderboard")
        for i, b in enumerate(bench_bests):
            if not isinstance(b, dict):
                raise SettingsModelDriverTabComposeError(
                    f"bench_bests[{i}] must be an object"
                )
            task = _require_nonempty(b.get("task"), field=f"bench_bests[{i}].task")
            best = _require_nonempty(
                b.get("best_model_id"), field=f"bench_bests[{i}].best_model_id"
            )
            score = b.get("score")
            if score is not None:
                if isinstance(score, bool) or not isinstance(score, (int, float)):
                    raise SettingsModelDriverTabComposeError(
                        f"bench_bests[{i}].score must be finite in [0, 1] when set"
                    )
                sf = float(score)
                if not math.isfinite(sf) or sf < 0 or sf > 1:
                    raise SettingsModelDriverTabComposeError(
                        f"bench_bests[{i}].score must be finite in [0, 1] when set"
                    )
            if focus is not None and task == focus:
                bench_best_for_focus = best
                bench_aligned = best == decision.selected_model_id
                notes.append(
                    f"bench_aligned=true for task={focus}"
                    if bench_aligned
                    else (
                        f"bench_aligned=false (selected={decision.selected_model_id} "
                        f"best={best})"
                    )
                )
        if focus is not None and bench_best_for_focus is None:
            notes.append(
                f"focus_task={focus} not in bench_bests — bench_aligned=null (no invent)"
            )
    elif focus_task is not None:
        notes.append("focus_task set without bench_bests — bench_aligned=null")

    nd_shadow_differs: bool | None = None
    nd_shadow_model: str | None = None
    if nd_shadow is not None:
        if not isinstance(nd_shadow, dict):
            raise SettingsModelDriverTabComposeError(
                "nd_shadow must be an object when set"
            )
        kill = nd_shadow.get("kill_switch_on")
        if not isinstance(kill, bool):
            raise SettingsModelDriverTabComposeError(
                "nd_shadow.kill_switch_on must be an explicit boolean"
            )
        rec = _require_nonempty(
            nd_shadow.get("recommended_model_id"),
            field="nd_shadow.recommended_model_id",
        )
        if kill:
            notes.append(
                "nd_shadow kill_switch_on=true — shadow suppressed (default off required)"
            )
            nd_shadow_differs = None
            nd_shadow_model = None
        else:
            nd_shadow_model = rec
            nd_shadow_differs = rec != decision.selected_model_id
            notes.append(
                f"nd_shadow_differs=true (shadow={rec}, selected={decision.selected_model_id}) — advisory only"
                if nd_shadow_differs
                else "nd_shadow_differs=false (shadow agrees with selected) — still not authority"
            )
        conf = nd_shadow.get("confidence")
        if conf is not None:
            if isinstance(conf, bool) or not isinstance(conf, (int, float)):
                raise SettingsModelDriverTabComposeError(
                    "nd_shadow.confidence must be finite in [0, 1] when set"
                )
            cf = float(conf)
            if not math.isfinite(cf) or cf < 0 or cf > 1:
                raise SettingsModelDriverTabComposeError(
                    "nd_shadow.confidence must be finite in [0, 1] when set"
                )

    tab_ready = inventory_count >= 1
    notes.append(
        "tab_ready=true — inventory present for operator selection"
        if tab_ready
        else "tab_ready=false — empty inventory"
    )
    notes.append("live_router_authorized=false")
    notes.append("secrets_stored=false")

    return SettingsModelDriverTabCompose(
        decision=decision,
        inventory_count=inventory_count,
        pending_add_count=pending_add_count,
        bench_aligned=bench_aligned,
        bench_best_for_focus=bench_best_for_focus,
        nd_shadow_differs=nd_shadow_differs,
        nd_shadow_model=nd_shadow_model,
        tab_ready=tab_ready,
        live_router_authorized=False,
        secrets_stored=False,
        notes=tuple(notes),
        authority="settings_model_driver_tab_compose_advisory",
    )


__all__ = [
    "SettingsModelDriverTabCompose",
    "SettingsModelDriverTabComposeError",
    "compose_settings_model_driver_tab",
]

File: apps/reading/src/api/settingsModelDriverTabCompose.ts
/**
 * Settings model driver tab compose (pure).
 *
 * Operator vision: decision-tree tab to select the AI model for a prompt,
 * with usage bar, budget projection, optional Antiek-bench weekly best,
 * and optional NotDiamond shadow recommendation (never authority).
 *
 * live_router_authorized always false.
 * secrets_stored always false (add-model is inventory-only here).
 */

import {
  composeModelDecisionWithProjection,
  type ModelDecisionPromptComposeInput,
  type ModelDecisionPromptComposeResult,
  type ModelOption,
} from "./modelDecisionPromptCompose";

export interface BenchTaskBest {
  /** Task family (e.g. deep_research, twin_notes). */
  task: string;
  /** Model id ranked best for task — caller-supplied only. */
  best_model_id: string;
  /** Optional score 0..1. */
  score?: number | null;
}

export interface NotDiamondShadowRec {
  /** ND recommended model id — advisory shadow only. */
  recommended_model_id: string;
  /** Kill switch must be off for ND to be considered shown. */
  kill_switch_on: boolean;
  confidence?: number | null;
}

export interface SettingsModelDriverTabInput {
  selected_model_id: string;
  models: ModelOption[];
  daily_cap_usd: number | null;
  spent_usd: number | null;
  projected_cost_usd_high?: number | null;
  projected_cost_usd_low?: number | null;
  /**
   * Optional Antiek-bench weekly bests (caller-supplied rankings only).
   * Never invents leaderboard rows.
   */
  bench_bests?: BenchTaskBest[] | null;
  /** Optional task to highlight for bench alignment. */
  focus_task?: string | null;
  /**
   * Optional NotDiamond shadow rec. Never becomes routing authority.
   */
  nd_shadow?: NotDiamondShadowRec | null;
  /**
   * Models pending add (BYOK inventory rows without secret material).
   * Secrets must not appear here — secrets_stored stays false.
   */
  pending_add_model_ids?: string[] | null;
}

export interface SettingsModelDriverTabCompose {
  decision: ModelDecisionPromptComposeResult;
  inventory_count: number;
  pending_add_count: number;
  /** True when selected model matches bench best for focus_task (if set). */
  bench_aligned: boolean | null;
  bench_best_for_focus: string | null;
  /** True when ND shadow recommends a different model and kill_switch off. */
  nd_shadow_differs: boolean | null;
  nd_shadow_model: string | null;
  /**
   * Tab ready for operator decision when inventory non-empty and decision
   * composed (would_exceed may still be null — honesty).
   */
  tab_ready: boolean;
  /** Always false — ND/local never auto-route from this surface. */
  live_router_authorized: false;
  /** Always false — no secret material accepted or stored. */
  secrets_stored: false;
  notes: string[];
  authority: "settings_model_driver_tab_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose the settings model-driver decision tree tab snapshot.
 * Never authorizes live routing; never stores secrets.
 */
export function composeSettingsModelDriverTab(
  input: SettingsModelDriverTabInput,
): SettingsModelDriverTabCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }

  const notes: string[] = [
    "live_router_authorized=false — operator selects model; no auto-router",
    "secrets_stored=false — inventory/ids only; never accept raw API keys here",
    "NotDiamond is advisory/shadow only (§16 REJECT as production router)",
  ];

  const decisionInput: ModelDecisionPromptComposeInput = {
    selected_model_id: input.selected_model_id,
    models: input.models,
    daily_cap_usd: input.daily_cap_usd,
    spent_usd: input.spent_usd,
    projected_cost_usd_high: input.projected_cost_usd_high,
    projected_cost_usd_low: input.projected_cost_usd_low,
  };
  const decision = composeModelDecisionWithProjection(decisionInput);
  notes.push(...decision.notes);

  const inventory_count = input.models.length;
  notes.push(`inventory_count=${inventory_count}`);

  let pending_add_count = 0;
  if (input.pending_add_model_ids != null) {
    if (!Array.isArray(input.pending_add_model_ids)) {
      throw new Error("pending_add_model_ids must be an array when set");
    }
    const seen = new Set<string>();
    for (let i = 0; i < input.pending_add_model_ids.length; i++) {
      const id = requireNonEmpty(
        input.pending_add_model_ids[i],
        `pending_add_model_ids[${i}]`,
      );
      // Reject anything that looks like a secret (fail closed).
      if (
        id.length > 128 ||
        /sk-|api[_-]?key|secret|bearer\s/i.test(id) ||
        id.includes(" ")
      ) {
        throw new Error(
          `pending_add_model_ids[${i}] must be a model id, not secret material`,
        );
      }
      if (seen.has(id)) {
        throw new Error(`duplicate pending_add_model_id: ${id}`);
      }
      seen.add(id);
    }
    pending_add_count = seen.size;
    notes.push(`pending_add_count=${pending_add_count} (ids only, no secrets)`);
  }

  let bench_aligned: boolean | null = null;
  let bench_best_for_focus: string | null = null;
  if (input.bench_bests != null) {
    if (!Array.isArray(input.bench_bests)) {
      throw new Error("bench_bests must be an array when set");
    }
    const focus =
      input.focus_task === undefined || input.focus_task === null
        ? null
        : requireNonEmpty(input.focus_task, "focus_task");
    if (input.bench_bests.length === 0) {
      notes.push("bench_bests empty — no invent leaderboard");
    }
    for (let i = 0; i < input.bench_bests.length; i++) {
      const b = input.bench_bests[i];
      if (!b || typeof b !== "object") {
        throw new Error(`bench_bests[${i}] must be an object`);
      }
      const task = requireNonEmpty(b.task, `bench_bests[${i}].task`);
      const best = requireNonEmpty(
        b.best_model_id,
        `bench_bests[${i}].best_model_id`,
      );
      if (b.score !== undefined && b.score !== null) {
        if (
          typeof b.score !== "number" ||
          !Number.isFinite(b.score) ||
          b.score < 0 ||
          b.score > 1
        ) {
          throw new Error(
            `bench_bests[${i}].score must be finite in [0, 1] when set`,
          );
        }
      }
      if (focus !== null && task === focus) {
        bench_best_for_focus = best;
        bench_aligned = best === decision.selected_model_id;
        notes.push(
          bench_aligned
            ? `bench_aligned=true for task=${focus}`
            : `bench_aligned=false (selected=${decision.selected_model_id} best=${best})`,
        );
      }
    }
    if (focus !== null && bench_best_for_focus === null) {
      notes.push(
        `focus_task=${focus} not in bench_bests — bench_aligned=null (no invent)`,
      );
    }
  } else if (input.focus_task != null) {
    notes.push("focus_task set without bench_bests — bench_aligned=null");
  }

  let nd_shadow_differs: boolean | null = null;
  let nd_shadow_model: string | null = null;
  if (input.nd_shadow != null) {
    if (typeof input.nd_shadow !== "object") {
      throw new Error("nd_shadow must be an object when set");
    }
    if (typeof input.nd_shadow.kill_switch_on !== "boolean") {
      throw new Error("nd_shadow.kill_switch_on must be an explicit boolean");
    }
    const rec = requireNonEmpty(
      input.nd_shadow.recommended_model_id,
      "nd_shadow.recommended_model_id",
    );
    if (input.nd_shadow.kill_switch_on) {
      notes.push(
        "nd_shadow kill_switch_on=true — shadow suppressed (default off required)",
      );
      nd_shadow_differs = null;
      nd_shadow_model = null;
    } else {
      nd_shadow_model = rec;
      nd_shadow_differs = rec !== decision.selected_model_id;
      notes.push(
        nd_shadow_differs
          ? `nd_shadow_differs=true (shadow=${rec}, selected=${decision.selected_model_id}) — advisory only`
          : `nd_shadow_differs=false (shadow agrees with selected) — still not authority`,
      );
    }
    if (
      input.nd_shadow.confidence !== undefined &&
      input.nd_shadow.confidence !== null
    ) {
      if (
        typeof input.nd_shadow.confidence !== "number" ||
        !Number.isFinite(input.nd_shadow.confidence) ||
        input.nd_shadow.confidence < 0 ||
        input.nd_shadow.confidence > 1
      ) {
        throw new Error(
          "nd_shadow.confidence must be finite in [0, 1] when set",
        );
      }
    }
  }

  const tab_ready = inventory_count >= 1;
  notes.push(
    tab_ready
      ? "tab_ready=true — inventory present for operator selection"
      : "tab_ready=false — empty inventory",
  );
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");

  return {
    decision,
    inventory_count,
    pending_add_count,
    bench_aligned,
    bench_best_for_focus,
    nd_shadow_differs,
    nd_shadow_model,
    tab_ready,
    live_router_authorized: false,
    secrets_stored: false,
    notes,
    authority: "settings_model_driver_tab_compose_advisory",
  };
}

export function formatSettingsModelDriverTabSummary(
  t: SettingsModelDriverTabCompose,
): string {
  const w =
    t.decision.would_exceed === null
      ? "would_exceed=null"
      : `would_exceed=${t.decision.would_exceed}`;
  return (
    `driver tab · model=${t.decision.selected_model_id} · ${w} · ` +
    `tab_ready=${t.tab_ready} · live_router_authorized=false · secrets_stored=false`
  );
}
